cpp/src/arrow/record_batch.cc

```diff
@@ -78,6 +79,52 @@ class SimpleRecordBatch : public RecordBatch {
 
   std::shared_ptr<ArrayData> column_data(int i) const override { return columns_[i]; }
 
+  Status AddColumn(int i, const std::shared_ptr<Field>& field,
+                   const std::shared_ptr<ArrayData>& column,
+                   std::shared_ptr<RecordBatch>* out) const override {
+    if (i < 0 || i > num_columns() + 1) {
```

Comment:
I think this should be `i > num_columns()`. This is also a bug in `SimpleTable::AddColumn`. Can you add a test where `i == num_columns()`?